### PR TITLE
Specify explicit public schema throughout

### PR DESCRIPTION
**Pull request by pmbohm**

### CODE REVIEW

**Comment by pmbohm** [comment]
Requires 
```
ALTER EXTENSION imos UPDATE to '1.1'; 

```
to be run

**Comment by lwgordonimos** [comment]
Just to get my vague thoughts down at this stage:

1. There are clearly pros and cons with forcing this to be in public schema. I personally haven't dug deep enough into these scenarios to know the best solution
1. The updating mechanism for extensions is separate from the original goal of "make dumps restorable if they contain materialised views", but obvious is relevant to making changes
1. I don't know how safe/desirable it is to set the `search_path` inside a function, I was just playing to see if it worked hypothetically

**Comment by jonescc** [comment]
So Tom Lane's advice (postgres developer/steering committee member) is

>The safe way to write SQL-language functions to be search-path-proof
is to schema-qualify the names in them, or to add a "SET search_path"
clause to the function definition.
>
>The problem with the latter approach is that it defeats inlining.
I thought for a minute that maybe we could teach the planner to do
inlining anyway by parsing the function body with the adjusted
search_path, but that doesn't really preserve the same semantics
(a SET would change the environment for called functions too).
>
>So for now, the recommendation has to be "write functions you want
to inline with schema qualifications".  If you're worried about
preserving relocatability of an extension containing such functions,
the @extschema@ feature might help.

So as discussed, I think we should stick with schema qualifying problem references.

Preferable to hacking the restore process to set schema paths or other fixups then I think.

**Comment by lwgordonimos** [comment]
> Preferable to hacking the restore process to set schema paths or other fixups then I think.

Agreed, that's the sustainable way forward I think, because in our case, I don't think being "relocatable" is important to us. 

The reason I was playing with search_path in function was for nested function calls, where qualifying the first function is not sufficient, because it then internally it calls other functions unqualified (and fails).

I was thinking that the only benefit I think to hacking the restore process to remove the `REFRESH` instruction is that *old* dumps (think monthly backups from last year) would be restorable.

**Comment by jonescc** [comment]
Should be OK to restore old backups if we're doing that into a db with the updated extensions.

So looking at the description of function inlining, I don't think it applies to us - it applies to functions which operate on scalar values and functions which returns tables.  The advantage of schema inlining (for table functions) is that the planner can push down constraints into the function definition. This can make a huge difference to performance where this applies.

I don't think this applies and specifying the search path (at the function level not in the body) is probably neater.

**Comment by lwgordonimos** [comment]
Sorry might be missing something.

Take an argo.dump from last year for example, it's still going to have in it a materialised view defined with unqualified function calls, e.g.

```sql
DECLARE
    result GEOMETRY;
BEGIN
    IF (accum is null) THEN
        RETURN point;
    ELSIF (point is null) THEN
        RETURN accum;
    ELSIF (st_geometrytype(accum) = 'ST_Point') THEN
        RETURN make_shortest_line(accum, point);
    ELSIF (st_geometrytype(accum) = 'ST_LineString') THEN
        RETURN add_point_shortest(accum, point);
    ELSIF (st_geometrytype(accum) = 'ST_MultiLineString') THEN
        SELECT INTO result st_geometryn(accum, 1);

        FOR idx in 2 .. st_numgeometries(accum) - 1 LOOP
            SELECT INTO result st_collect(result, st_geometryn(accum, idx));
        END LOOP;

        RETURN st_collectionextract(st_collect(result, add_point_shortest(st_geometryn(accum, st_numgeometries(accum)), point)),2);
    ELSE
        RETURN null;
    END IF;
END;
```

Without setting a search_path (i.e. some kind of restore hack), there's nothing we can do to make *this* dump file restorable, regardless of any changes to the extension?

**Comment by jonescc** [comment]
Might be my misunderstanding - I haven't seen any suggestions to change the argo harvester to fix issues with the restoring the db.  If that's required then yes to restore old schemas we'll need to do the search_path hack.

**Comment by lwgordonimos** [comment]
Sorry I pasted the wrong thing. Now I see where I'm getting confused.

```sql
set search_path = 'argo';
refresh materialized view argo_trajectory_map;
```
This fails with the same error as pg_restore, but what's confused me is how it finds the first function, `make_trajectory`, without public in the search path. It fails on the nested functions but not the *directly* called function.

So even though the search path doesn't include public, it still finds `make_trajectory` from the public schema when refreshing the view (but then fails because of inner functions). That's the bit that confused me!

But if I add the `search_path` to the function itself (not in body, which also worked but was incorrect as you pointed out!), it does work despite public *not* being in the search_path.

```sql
CREATE OR REPLACE FUNCTION make_trajectory(accum geometry, point geometry)
RETURNS geometry AS
$BODY$ DECLARE
    result GEOMETRY;
BEGIN
    IF (accum is null) THEN
        RETURN point;
    ELSIF (point is null) THEN
        RETURN accum;
    ELSIF (st_geometrytype(accum) = 'ST_Point') THEN
        RETURN make_shortest_line(accum, point);
    ELSIF (st_geometrytype(accum) = 'ST_LineString') THEN
        RETURN add_point_shortest(accum, point);
    ELSIF (st_geometrytype(accum) = 'ST_MultiLineString') THEN
        SELECT INTO result st_geometryn(accum, 1);

        FOR idx in 2 .. st_numgeometries(accum) - 1 LOOP
            SELECT INTO result st_collect(result, st_geometryn(accum, idx));
        END LOOP;

        RETURN st_collectionextract(st_collect(result, add_point_shortest(st_geometryn(accum, st_numgeometries(accum)), point)),2);
    ELSE
        RETURN null;
    END IF;
END;
$BODY$
LANGUAGE plpgsql VOLATILE
SET search_path = 'public'
COST 100;
```

So the million dollar question for me is how this *doesn't* fail with an error about not being able to find `make_trajectory`.

```sql
set search_path = 'argo';
refresh materialized view argo_trajectory_map;
```

**Comment by pmbohm** [comment]
One option for testing is to use this change in 2-aws-syd.emii.org.au.json
```
"git_extension_support": [
                    {
                        "name": "imos",
                        "branch": "backlog-1680",
                        "repository": "https://github.com/aodn/harvest_sql_library"
                    }
                ]
```

**Comment by lwgordonimos** [comment]
Ok *now* I see what this PR is doing. It defines a migration from 1.0 -> 1.1 in addition to the new complete definitions. Seems neat.

As I mentioned above, it seems like it does actually work (haven't tested with pg_restore, only in pgadmin), but I still don't know why it works!

**Comment by jonescc** [comment]
I'm kind of assuming that its using extension info in pg_catalog (always included in the serach oath even if you don't specify it).  But I haven't had a chance to find out why.  Its implied in all the references above.

Happy to test in vagrant - ran up a vagrant instance this morning for that purpose.

**Comment by lwgordonimos** [comment]
On my local instance, I don't have it installed as an extension, I must have just pasted in the extension SQL at some point manually.

Despite that, it still finds all of the functions anyway, so I think you're right it's just some implicit thing that finds functions independently of the search_path, but also seems to be unrelated to the extensions mechanism as well.

All I can come up with is that the materialized view stores the search_path used when creating it, which allows the refresh to occur when public is not in the active search_path when initiating the refresh:

```sql
-- public must be in the path when the view is created, otherwise "function database_activity() does not exist"
set search_path = argo,public;

-- define a simple view which depends on a function from public schema
drop materialized view argo.db_activity;
create materialized view argo.db_activity as
    select database_activity()
with no data;

-- removes public from search path, but can still refresh successfully
set search_path = argo;
refresh materialized view argo.db_activity;

-- succeeds
select * from argo.db_activity;

-- fails - "function database_activity() does not exist"
select database_activity()
```

**Comment by jonescc** [comment]
All works fine in vagrant - alter extension worked when argo schema populated with a previous hacked restore.  Dropped the schema and then reprovisioned and could then happily restore the argo dump without hacks.

We can test the restores using a db with the extension updates.  And should probably apply extension updates in RC and perhaps leave for a while to ensure there are no issues before applying in prod.

Any reason not to merge this?  Requires changes to be applied manually in RC/Prod.

**Comment by lwgordonimos** [comment]
Looks good to me, but man it needs those commits squashed ;)

**Comment by pmbohm** [comment]
@lwgordonimos 
If you pasted in the SQL that creates the functions, they should be there in the public repo.
What if there was some data in the Materialised view? Does it all goes to poop

**Comment by lwgordonimos** [comment]
@pmbohm sorry I don't quite follow. Yes the functions are there, and the view does get updated with the expected content in that contrived case. Is that what you meant?

File: extension/imos--unpackaged--1.0.sql [hunk 1]
```diff
@@ -4,24 +4,24 @@
 \echo Use "CREATE EXTENSION imos" to load this file. \quit
 
 -- Harvester
-ALTER EXTENSION "imos" ADD function getendtoken();
-ALTER EXTENSION "imos" ADD function format_name();
-ALTER EXTENSION "imos" ADD function add_point_shortest();
-ALTER EXTENSION "imos" ADD function make_trajectory();
-ALTER EXTENSION "imos" ADD function make_shortest_line();
+ALTER EXTENSION "imos" ADD function public.getendtoken();
```
Comment (lwgordonimos) [inline extension/imos--unpackaged--1.0.sql:7]: What are these changes doing? I thought unpackaged extensions were a legacy thing we didn't have to deal with any more?

Comment (jonescc) [reply extension/imos--unpackaged--1.0.sql:7]: yes, these changes aren't necessary, but I was letting that one slide.  the unpackaged--1.0.sql file is for adding the unmanaged functions to the extension in the first place.  that was possibly used a long time ago to add functions probably managed by an sql script at the time to the extension to get us to version 1.0

**Comment by jonescc** [comment]
We're going to have to test this against the current state of the extension in prod to ensure this will actually work there.  Because the chef extension management was never fixed, the current state of prod  probably doesn't match the current 1.0 extension.    I suspect we'll need to fix that up manually so we can start using this mechanism properly.

**Comment by jonescc** [comment]
I for one would prefer to use the alter extension mechanism in future with its manual step than hacking around in prod as has been the approach up until now.

**Comment by lwgordonimos** [comment]
> I for one would prefer to use the alter extension mechanism in future with its manual step than hacking around in prod as has been the approach up until now until such time as we revisit this.

Yeah definitely. I had no idea the extensions could natively support migrations like this, so I'm all for it. Much more testable.

**Pull request merged**